penguin/system/runtime_event_ledger.py

```
@@ -0,0 +1,687 @@
+                    clauses.append("(agent_id = ? OR agent_id IS NULL)")
+                    params.append(agent_id)
+                if directory:
+                    clauses.append("(directory = ? OR directory IS NULL)")
+                    params.append(directory)
+
+                sql = """
+                    SELECT event_json
+                    FROM runtime_events
+                    WHERE {where_clause}
+                    ORDER BY rowid ASC
+                """.format(where_clause=" AND ".join(clauses))
```

Ruff (0.15.20)

[error] Possible SQL injection vector through string-based query construction

(S608)